File: docker-convert.js
const fs = require('fs/promises'); // Importar módulo para manejar archivos (promesas)
const { exec, execSync } = require('child_process'); // Importar métodos necesarios para ejecutar comandos
const process = require('process'); // Importar módulo para acceder a argumentos
const util = require('util');

// Definir variables constantes
const dockerfileKeywords = [
    'FROM', 'MAINTAINER', 'RUN', 'CMD', 'LABEL', 'EXPOSE', 'ENV', 'ADD', 'COPY',
    'ENTRYPOINT', 'VOLUME', 'USER', 'WORKDIR', 'ARG', 'ONBUILD', 'STOPSIGNAL',
    'HEALTHCHECK', 'SHELL'
];

const regexCreatedBy = /CREATED BY\s+(.+)/;

// Convierte exec a una función que devuelve una promesa
const execPromise = util.promisify(exec);

/**
 * Verifica si una línea comienza con una palabra clave de Dockerfile
 * @param {string} line - La línea a verificar
 * @returns {boolean} - Retorna true si la línea comienza con una palabra clave, false de lo contrario
 */
function isFilteredLine(line) {
    return dockerfileKeywords.some(keyword => line.startsWith(keyword));
}

/**
 * Extrae texto después de "CREATED BY" en una línea
 * @param {string} line - La línea de texto
 * @returns {string|null} - Retorna el texto después de "CREATED BY" o null si no hay coincidencia
 */
function getTextoDespuesDeCreatedBy(line) {
    const match = regexCreatedBy.exec(line);
    return match ? match[1].trim() : null;
}

/**
 * Guarda contenido en un archivo Dockerfile
 * @param {string} contenido - El contenido a guardar en el archivo
 * @param {string} nombre - El nombre base del archivo Dockerfile
 */
async function guardarEnDockerfile(contenido, nombre) {
    let nombreArchivo = `Dockerfile-${nombre}`;
    let i = 0;

    while (true) {
        try {
            await fs.writeFile(nombreArchivo, contenido);
            console.log(`Archivo '${nombreArchivo}' guardado exitosamente.`);
            break;
        } catch (err) {
            if (err.code === 'EEXIST') {
                const respuesta = await prompt(`El archivo '${nombreArchivo}' ya existe. ¿Desea reemplazarlo? (s/n): `);
                if (respuesta.toLowerCase() === 's') {
                    await fs.writeFile(nombreArchivo, contenido);
                    console.log(`Archivo '${nombreArchivo}' reemplazado exitosamente.`);
                    break;
                } else if (respuesta.toLowerCase() === 'n') {
                    i++;
                    nombreArchivo = `Dockerfile-${i}-${nombre}`;
                } else {
                    console.error('Respuesta no válida. Por favor, ingrese "s" para sí o "n" para no.');
                }
            } else {
                console.error(`Error al guardar el archivo: ${err.message}`);
                return;
            }
        }
    }
}

/**
 * Analiza el historial de Docker y extrae instrucciones relevantes
 * @param {string} history - El historial de Docker como una cadena
 * @returns {string[]} - Retorna un arreglo de instrucciones Dockerfile extraídas del historial
 */
function parseHistoryFile(history) {
    if (typeof history !== 'string') {
        console.error('La variable history no es una cadena, es %s', JSON.stringify(history));
        return [];
    }

    const lines = history.split('\n');
    const instructions = [];
    const idx = lines[0].indexOf('CREATED BY');
    const last = lines[0].indexOf('SIZE');

    for (const line of lines) {
        const instruction = line.substring(idx, last).trim() + '\n';
        if (isFilteredLine(instruction)) {
            instructions.push(instruction);
        }
    }

    return instructions.reverse();
}

/**
 * Genera el contenido de un Dockerfile
 * @param {string[]} instructions - Las instrucciones Dockerfile
 * @param {string|null} baseImage - La imagen base opcional
 * @returns {string} - El contenido generado del Dockerfile
 */
function generateDockerfile(instructions, baseImage = null) {
    let dockerfileContent = '';

    if (baseImage) {
        dockerfileContent += `FROM ${baseImage}\n`;
    }

    for (const instruction of instructions) {
        dockerfileContent += instruction;
    }

    return dockerfileContent;
}

/**
 * Obtiene el historial de una imagen Docker
 * @param {string} imageName - El nombre de la imagen Docker
 * @returns {Promise<string>} - Promesa que retorna el historial de la imagen Docker
 */
async function obtenerDockerHistory(imageName) {
    const command = `docker history --no-trunc ${imageName}`;
    console.log("Comando de historial: ", command);

    try {
        const { stdout, stderr } = await execPromise(command);

        if (stderr) {
            console.error(`Error al ejecutar el comando: ${stderr}`);
            return;
        }

        return stdout;
    } catch (error) {
        console.error(`Error al ejecutar el comando: ${error.message}`);
    }
}

/**
 * Lista las imágenes Docker disponibles
 * @returns {string[]} - Arreglo de líneas de la salida del comando 'docker images'
 */
async function listarImagenesDocker() {
    const command = 'docker images';

    try {
        const { error, stdout, stderr } = await execPromise(command);
        if (!stdout) {
            console.error('El número de imagen debe ser mayor que cero.');
        } else {
            return stdout;
        }
    } catch (err) {
        console.error(`Error al listar las imágenes Docker: ${err}: ${stderr}, ${error}`);
    }
}

/**
 * Solicita al usuario el número de una imagen Docker
 * @returns {Promise<number>} - Promesa que retorna el número de imagen ingresado por el usuario
 */
async function solicitarNumeroImagen() {
    while (true) {
        try {
            const numeroImagen = parseInt(await prompt('Ingrese el número de la imagen que desea solicitar: '));
            if (numeroImagen <= 0 && numeroImagen > top) {
                console.error('El número de imagen debe ser mayor que cero.');
            } else {
                return numeroImagen;
            }
        } catch (err) {
            console.error('Valor no válido. Por favor, ingrese un número entero.');
        }
    }
}

/**
 * Muestra un prompt al usuario y retorna la respuesta
 * @param {string} message - El mensaje a mostrar en el prompt
 * @returns {Promise<string>} - Promesa que retorna la respuesta del usuario
 */
function prompt(message) {
    const rl = require('readline').createInterface({
        input: process.stdin,
        output: process.stdout
    });

    return new Promise((resolve) => {
        rl.question(message, (answer) => {
            rl.close();
            resolve(answer.trim());
        });
    });
}

/**
 * Función principal del script
 * Realiza todo el flujo de trabajo para obtener el historial de Docker, analizarlo, y generar un Dockerfile
 */
async function start() {
    try {
        // Obtener la imagen base (de argumento o selección por usuario)
        let baseImage = process.argv[2] || null;

        if (!baseImage) {
            const imagenes = (await listarImagenesDocker()).toString();
            console.log(imagenes);
            //console.log(imagenes.substring(imagenes.indexOf("IMAGE ID"), imagenes.indexOf("CREATED")));
            const numeroImagen = await solicitarNumeroImagen();
            console.log("numero imagen %i, length %s, primer barra-n %s, cantidad de elementos %s", numeroImagen, imagenes.length, imagenes.indexOf('\n'), imagenes.toString().split('\n').length);
            //console.log("primer imagen %s", imagenes.toString().split('\n')[0]);
            const imagen = imagenes.split('\n')[numeroImagen];
            baseImage = imagen.substring(imagenes.indexOf('IMAGE ID'), imagenes.indexOf('CREATED')).trim();
            console.log(`\n[${baseImage}]\n`);
        }

        // Obtener el historial de Docker
        console.log("Obtener el historial de Docker " + baseImage);
        const history = await obtenerDockerHistory(baseImage);

        // Analizar el historial y generar el Dockerfile
        const instructions = parseHistoryFile(history);
        const dockerfileContent = generateDockerfile(instructions, baseImage);

        // Imprimir y guardar el Dockerfile
        console.log(dockerfileContent);
        await guardarEnDockerfile(dockerfileContent, baseImage);
    } catch (err) {
        console.error(err);
    }
}

// Punto de entrada del script
(async () => {
    try {
        await start();
    } catch (err) {
        if (err.name === 'KeyboardInterrupt') {
            console.log('\nScript interrumpido por el usuario.');
            // Otra acción que desees realizar al interrumpir el script
            process.exit(0);
        } else {
            console.error(err);
        }
    }
})();

File: docker-convert.py
# -*- coding: utf-8 -*-
import sys
import subprocess
import re

dockerfile_keywords = [
    "FROM",
    "MAINTAINER",
    "RUN",
    "CMD",
    "LABEL",
    "EXPOSE",
    "ENV",
    "ADD",
    "COPY",
    "ENTRYPOINT",
    "VOLUME",
    "USER",
    "WORKDIR",
    "ARG",
    "ONBUILD",
    "STOPSIGNAL",
    "HEALTHCHECK",
    "SHELL"
]

def is_filtered_line(line):
    #_txt = filter(lambda x: text.find(x) != -1, dockerfile_keywords)
    #idx 
    return line if any(line.find(keyword) == 0 for keyword in dockerfile_keywords) else ""

instructions = []

# # Ejemplo de uso:
# linea_de_texto = "IMAGE CREATED              CREATED BY"
# texto_despues_created_by = obtener_texto_despues_de_created_by(linea_de_texto)
# print("Texto después de 'CREATED BY':", texto_despues_created_by)
def obtener_texto_despues_de_created_by(linea):
    # Buscar el índice de "CREATED BY" en la línea de texto
    idx = linea.find("CREATED BY")

    # Si no se encuentra "CREATED BY", retornar None
    if idx == -1:
        print('nada nada')
        return None

    # Tomar el texto que comienza después de "CREATED BY"
    texto_despues_created_by = linea[idx + len("CREATED BY"):].strip()

    return texto_despues_created_by

def guardar_en_dockerfile(contenido, name):
    # Nombre del archivo
    nombre_archivo = f"Dockerfile-{name}"
    i = 0

    # Intentar guardar en el archivo Dockerfile
    while True:
        try:
            if i > 0:
                # Intentar con nombres como Dockerfile0, Dockerfile1, etc.
                nombre_archivo = f"Dockerfile{i}-{name}"
            # Intentar abrir el archivo para escribir
            with open(nombre_archivo, "x") as archivo:
                archivo.write(contenido)
            print(f"Archivo '{nombre_archivo}' guardado exitosamente.")
            break  # Salir del bucle si se pudo guardar correctamente
        except FileExistsError:
            # Si el archivo ya existe
            respuesta = input(f"El archivo '{nombre_archivo}' ya existe. ¿Desea reemplazarlo? (s/n): ")
            if respuesta.lower() == 's':
                # Intentar abrir el archivo para escribir (sobrescribir)
                with open(nombre_archivo, "w") as archivo:
                    archivo.write(contenido)
                print(f"Archivo '{nombre_archivo}' reemplazado exitosamente.")
                break  # Salir del bucle si se pudo reemplazar correctamente
            elif respuesta.lower() == 'n':
                # Incrementar el índice para intentar con otro nombre
                i += 1
            else:
                print("Respuesta no válida. Por favor, ingrese 's' para sí o 'n' para no.")

def parse_history_file(filename):
    lines = filename.split("\n")
    instructions = []
    idx = lines[0].find("CREATED BY")
    last = lines[0].find("SIZE")
    
    # with open(filename, "r") as f:
    #     for line in f:
    #         if idx == -1:
    #             idx = line.find("CREATED BY")
    #             print(idx)
    #             last = line.find("SIZE")
    #             print(last)
    #             break

    # with open(filename, "r") as f:
        #for line in f:
    for line in lines:
        instruction = line[idx:last].strip() + "\n"
        if(is_filtered_line(instruction)):
            instructions.append(instruction)

    return instructions[::-1]

def generate_dockerfile(instructions, base_image=None):
    dockerfile_content = ""

    if base_image:
        dockerfile_content = f"FROM {base_image}" + "\n"

    for instruction in instructions:
        dockerfile_content += f"{instruction}"

    return dockerfile_content

import subprocess

def obtener_docker_history(nombre_imagen):
    # Ejecutar el comando 'docker history' y capturar la salida
    proceso = subprocess.Popen(['docker', 'history', '--no-trunc', nombre_imagen], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    salida, error = proceso.communicate()

    # Decodificar la salida a una cadena de texto
    salida_decodificada = salida.decode('utf-8')

    # Retornar la salida y el posible error
    return salida_decodificada, error

def listar_imagenes_docker():
    try:
        # Ejecutar el comando 'docker images' y capturar la salida
        proceso = subprocess.Popen(['docker', 'images'], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        salida, error = proceso.communicate()

        # Decodificar la salida a una cadena de texto
        salida_decodificada = salida.decode('utf-8')

        # Separar la salida por líneas y guardarla en un array
        lineas = salida_decodificada.split('\n')

        # Imprimir la salida numerada
        print("Listado de imágenes Docker:")
        for i, linea in enumerate(lineas):
            idx = f"{i} ." if i > 0 else "   "
            print(f"{idx}{linea}")
        
        return lineas
    except Exception as e:
        print("Error al listar las imágenes Docker:", e)

def solicitar_numero_imagen():
    while True:
        try:
            numero_imagen = int(input("Ingrese el número de la imagen que desea solicitar: "))
            if numero_imagen <= 0:
                print("El número de imagen debe ser mayor que cero.")
            else:
                return numero_imagen
        except ValueError:
            print("Por favor, ingrese un número válido.")

def start():
    try:
        # Obtener el nombre de la imagen del argumento de la línea de comandos
        base_image = sys.argv[1]
    except:
        imagenes = listar_imagenes_docker()
        numero_imagen = solicitar_numero_imagen()
        imagen = imagenes[numero_imagen]
        base_image = imagen[imagenes[0].find("IMAGE"):imagenes[0].find("CREATED")].strip()
        print(f"[{base_image}] \n")
        
    # Obtener el historial de Docker y almacenar la salida en una variable
    historial, error = obtener_docker_history(base_image)

    # Imprimir la salida y manejar cualquier error
    if error:
        print("Error al obtener el historial de Docker:", error)

    instructions = parse_history_file(historial)
    dockerfile_content = generate_dockerfile(instructions, base_image)
    
    print(dockerfile_content)
    guardar_en_dockerfile(dockerfile_content, name= base_image)

if __name__ == "__main__":
    try:
        start()
    except KeyboardInterrupt:
        print("\nScript interrumpido por el usuario.")
        # Otra acción que desees realizar al interrumpir el script
        sys.exit(0)